doc/developer/design/20230223_stabilize_with_mutually_recursive.md

```diff
@@ -0,0 +1,297 @@
+* Use the internal observability metrics as early adopters for WMR.
+
+Validating (2) is [an open question](#unresolved-questions).
+
+## Testing and observability
+[testing-and-observability]: #testing-and-observability
+<!--
+Testability and explainability are top-tier design concerns!
+Describe how you will test and roll out the implementation.
+When the deliverable is a refactoring, the existing tests may be sufficient.
+When the deliverable is a new feature, new tests are imperative.
+
+Describe what metrics can be used to monitor and observe the feature.
+What information do we need to expose internally, and what information is interesting to the user?
+How do we expose the information?
+
+Basic guidelines:
+
+* Nearly every feature requires either Rust unit tests, sqllogictest tests, or testdrive tests.
+* Features that interact with Kubernetes additionally need a cloudtest test.
+* Features that interact with external systems additionally should be tested manually in a staging environment.
+* Features or changes to performance-critical parts of the system should be load tested.
+-->
+
+We plan to build up confidence in the updated query optimization pipeline by adding new tests and revisiting existing tests.
+Test scenarios can be categorized along two dimensions:
+
+### By type
+
+1. **Unit tests.**
+We aim to have one unit test per transform.
+We can invest time proportional to the complexity of the transform to ensure that each transform is correct.
+
+1. **Integration tests.**
+We will add test scenarios inspired by the use cases of our prospects as end-to-end `*.slt` tests.
+We will also add at least one long-running `mzcompose` test runs [as part of our nightly tests](https://buildkite.com/materialize/nightlies) and is used when qualifying future releases.
+
+As those tests will include expected results, it will be great if we have a reference external iteration driver for the semantics proposed in the original design doc[^wmr].
+That way we can cross-check the results of the reference and the internal implementation of WMR support and ensure that both produce equal results.
+We can implement such driver in Python and integrate it in our `mzcompose` tests.
+
+### By test scenario
+
+1. **Synthetic tests.**
+The best synthetic use case that we have identified so far seems to be the LDBC social network benchmark[^ldbc].
+Although there is a dedicated epic for that (MaterializeInc/materialize#17591), we can a selected subset of the work in order to bootstrap a testing environment that consists of
+  (a) LDBC data + updates, and
+  (b) several of the recursive queries defined by the benchmark.
+We can use the choke-point characterization of each query to figure out the most representative subset.
+
+1. **Use-case driven.**
+It is unclear how useful these will be as load tests, as we don't have the resources to write realistic data generators to replicate the domain of specific customers.
```

Comment:
@heeringa / @sjwiesman maybe we can discuss (possibly offline) possible ways to approach customers that are willing to act as beta testers for the feature. I think 2-3 use cases should be enough for for the epic, and AFAICT some of those can be also motivated by external use cases.